src/EnergyPlus/DataHeatBalance.cc

```diff
@@ -400,29 +400,27 @@ void CheckAndSetConstructionProperties(EnergyPlusData &state,
-             thisMaterial->Group == Material::MaterialGroup::WindowBlind || thisMaterial->Group == Material::MaterialGroup::Screen ||
-             thisMaterial->Group == Material::MaterialGroup::WindowSimpleGlazing ||
-             thisMaterial->Group == Material::MaterialGroup::ComplexWindowShade || thisMaterial->Group == Material::MaterialGroup::ComplexWindowGap ||
-             thisMaterial->Group == Material::MaterialGroup::GlassEquivalentLayer ||
-             thisMaterial->Group == Material::MaterialGroup::ShadeEquivalentLayer ||
-             thisMaterial->Group == Material::MaterialGroup::DrapeEquivalentLayer ||
-             thisMaterial->Group == Material::MaterialGroup::ScreenEquivalentLayer ||
-             thisMaterial->Group == Material::MaterialGroup::BlindEquivalentLayer ||
-             thisMaterial->Group == Material::MaterialGroup::GapEquivalentLayer);
-        bool TypeIsNotWindow =
-            (thisMaterial->Group == Material::MaterialGroup::Invalid || thisMaterial->Group == Material::MaterialGroup::Air ||
-             thisMaterial->Group == Material::MaterialGroup::RegularMaterial || thisMaterial->Group == Material::MaterialGroup::EcoRoof ||
-             thisMaterial->Group == Material::MaterialGroup::IRTMaterial);
+            (thisMaterial->group == Material::Group::WindowGlass || thisMaterial->group == Material::Group::WindowGas ||
+             thisMaterial->group == Material::Group::WindowGasMixture || thisMaterial->group == Material::Group::Shade ||
+             thisMaterial->group == Material::Group::WindowBlind || thisMaterial->group == Material::Group::Screen ||
+             thisMaterial->group == Material::Group::WindowSimpleGlazing || thisMaterial->group == Material::Group::ComplexWindowShade ||
+             thisMaterial->group == Material::Group::ComplexWindowGap || thisMaterial->group == Material::Group::GlassEquivalentLayer ||
+             thisMaterial->group == Material::Group::ShadeEquivalentLayer || thisMaterial->group == Material::Group::DrapeEquivalentLayer ||
+             thisMaterial->group == Material::Group::ScreenEquivalentLayer || thisMaterial->group == Material::Group::BlindEquivalentLayer ||
+             thisMaterial->group == Material::Group::GapEquivalentLayer);
+        bool TypeIsNotWindow = (thisMaterial->group == Material::Group::Invalid || thisMaterial->group == Material::Group::Air ||
+                                thisMaterial->group == Material::Group::Regular || thisMaterial->group == Material::Group::EcoRoof ||
+                                thisMaterial->group == Material::Group::IRTransparent);
         if (!thisConstruct.TypeIsWindow && !TypeIsNotWindow) assert(false);
     }
 
     if (InsideMaterNum == 0) return;
+    auto const *thisMaterialInside = dynamic_cast<const Material::MaterialChild *>(state.dataMaterial->Material(InsideMaterNum));
```

Comment:
All this dynamic casting makes me think that a next iteration of this work should do something like this:

```c++
struct MaterialBase {
  std::string name;
  virtual void applyProperty() {} // assuming the base class doesn't need to do anything
};

struct MaterialDerived : MaterialBase {
  Real64 someDerivedProperty;
  virtual void applyProperty() override;
};
```

then in the cc file:

```c++
void MaterialDerived::applyProperty() {
  this->something = someDerivedProperty ** 2;
}
```

This way you can just call `applyProperty` on the instance, and it will automatically use the right version regardless of which type of actual material you have.

Reply:
Does this mean when we want to get a member of a struct, we'll need to have a get and a set virtual function for every member of every derived struct?

Reply:
It means we will want to think about what we actually need the material classes to do/provide.  Class inheritance structures should provide layers of abstraction.  I'm not proposing we want more than one layer of inheritance here, but even in this one layer, we should still find ways to abstract the functionality and data.  Let's say you need to calculate F for a material.  In one material, F is a function of 2 material properties: x and y.   In another material, F is a function of 3 material properties: x, y, and z.  Out in client code, you don't want to constantly dynamic cast instances to each derived type and grab individual material properties and using them to calculate F.  No, you want to create a pure virtual function to calculate F in the base class.   Then each derived material type must implement their own function to calculate F.  And if x and y are actually shared by all the material types, those should be stored in the base class as well, and each derived type only needs to add data specific to their type.  This tightens up the interface between the material class and the clients using it and encourages code reuse, data hiding, and proper ownership.

This PR is a great step forward for sure.  Most of what I am saying here is just me being optimistic about where this will go next.  Thanks @yujiex